### ISSUE TITLE
Too many open files when querying security groups

### ISSUE THREAD

**Issue by vlerenc** [cloudfoundry/cli#878]
Hi,

when running `cf security-groups` on a system with many security-groups (>2K, don't know the exact limit when it fails), we get the error below. Since we use security-groups to control access to service instances and we have many spaces and especially service instances in our landscapes, this is a problem for us. Are you aware of this or are security groups not meant to be used so fine-grained, but how else then should we ensure that users can't access service instances in other spaces (besides the password of course; security groups and their resulting iptables rules seemed a perfect additional security measure).

Cheers, Vedran

### Command

```
cf security-groups
```

### CLI Version

```
cf version 6.19.0+b29b4e0-2016-06-08
```

### Error

```
Getting security groups as admin
FAILED
Error performing request: Get https://api.cf.service-fabrik.sc6.sapcloud.io/v2/security_groups/a7fbbcde-d2ff-4ad8-8721-ee0398c19be1/spaces?inline-relations-depth=1: dial tcp: lookup api.cf.service-fabrik.sc6.sapcloud.io on 172.18.4.24:53: dial udp 172.18.4.24:53: socket: too many open files
TIP: If you are behind a firewall and require an HTTP proxy, verify the https_proxy environment variable is correctly set. Else, check your network connection.
```

### Stack Trace

None given.

### CC API Endpoint Version

```
API endpoint: https://api.cf.service-fabrik.sc6.sapcloud.io (API version: 2.55.0)
```

### CF Trace

Too long, 13 MB. Some "interessting" sections:

```
RESPONSE: [2016-07-02T11:37:01Z]
HTTP/1.1 200 OK^M
Content-Length: 46632^M
Content-Type: application/json;charset=utf-8^M
Date: Sat, 02 Jul 2016 11:37:01 GMT^M
Server: nginx^M
X-Content-Type-Options: nosniff^M
X-Vcap-Request-Id: 017d675f-81eb-47f8-5bc5-05e80f4e8073^M
X-Vcap-Request-Id: 9abd8fd8-43fd-4818-608c-d58b03015ca0::db9f973c-015d-4c00-a8f8-66a10e84fc35^M
^M
{
  "total_results": 2004,
  "total_pages": 41,
  "prev_url": null,
  "next_url": "/v2/security_groups?order-direction=asc&page=2&results-per-page=50",
  "resources": [
```

Results for many spaces are reported I believe, but then only the requests (many) are found:

```
REQUEST: [2016-07-02T11:37:30Z]
GET /v2/security_groups/70437608-a655-4400-9cfa-83e7929b15e7/spaces?inline-relations-depth=1 HTTP/1.1^M
Host: api.cf.service-fabrik.sc6.sapcloud.io^M
Accept: application/json^M
Authorization: [PRIVATE DATA HIDDEN]
Content-Type: application/json^M
User-Agent: go-cli 6.19.0+b29b4e0 / linux^M
^M


REQUEST: [2016-07-02T11:37:30Z]
GET /v2/security_groups/0a647312-b263-42e3-900f-c59e5c036764/spaces?inline-relations-depth=1 HTTP/1.1^M
Host: api.cf.service-fabrik.sc6.sapcloud.io^M
Accept: application/json^M
Authorization: [PRIVATE DATA HIDDEN]
Content-Type: application/json^M
User-Agent: go-cli 6.19.0+b29b4e0 / linux^M
^M
```

And then it ends with:

```
REQUEST: [2016-07-02T11:37:32Z]
GET /v2/security_groups/a7fbbcde-d2ff-4ad8-8721-ee0398c19be1/spaces?inline-relations-depth=1 HTTP/1.1^M
Host: api.cf.service-fabrik.sc6.sapcloud.io^M
Accept: application/json^M
Authorization: [PRIVATE DATA HIDDEN]
Content-Type: application/json^M
User-Agent: go-cli 6.19.0+b29b4e0 / linux^M
^M

FAILED
Error performing request: Get https://api.cf.service-fabrik.sc6.sapcloud.io/v2/security_groups/a7fbbcde-d2ff-4ad8-8721-ee0398c19be1/spaces?inline-relations-depth=1: dial tcp: lookup api.cf.service-fabrik.sc6.sapcloud.io on 172.18.4.24:53: dial udp 172.18.4.24:53: socket: too many open files
TIP: If you are behind a firewall and require an HTTP proxy, verify the https_proxy environment variable is correctly set. Else, check your network connection.
```

### Platform Details

```
Distributor ID: Ubuntu
Description:    Ubuntu 14.04.4 LTS
Release:        14.04
Codename:       trusty
```

### Shell

Terminal/Bash

**Comment by dkoper** [comment]
Thanks @vlerenc, we'll take a look!

**Comment by vlerenc** [comment]
@dkoper Great, thank you as it's a significant component in our security concept and we'd like to keep it in as an additional means of isolation and to secure the landscapes from malicious attacks.

**Comment by dkoper** [comment]
Hi @vlerenc 

We believe we fixed this. Would you mind downloading an edge build and report back whether the issue is addressed for you?

Cheers,
Dies Koper
CF CLI PM

**Comment by vlerenc** [comment]
@dkoper Sure, np. How/from where?

**Comment by dkoper** [comment]
https://github.com/cloudfoundry/cli/blob/master/README.md#downloads
Check out the second last paragraph for links to our edge builds.

**Comment by vlerenc** [comment]
Yes, your edge binary is working, thank you. However, it takes quite some time (even though I am executing it from an inception box in the same OpenStack tenant/IaaS).

```
$ time ~/tmp/cf security-groups                                                                                                                                                                                                                                                                                       
Getting security groups as admin
OK

        Name                                                  Organization   Space
#0      public_networks
#1      dns
...
#2006

real    1m1.388s
user    0m9.974s
sys     0m1.006s
```

Is this the data transfer or can this be improved from performance PoV?

In any case, thank you for the fix already preventing the CLI to crash.

**Comment by XenoPhex** [comment]
At the moment, we make a list request for security groups (in series) and then do a spaces list request for each security group (in series). This means if you have 2006 security groups, we're doing *at least* 2006/50 + 2006 requests to generate that table.

Right now we have no plans on making these requests in parallel, which would really speed things up. But if this is a significant issue for you, please feel free to file a separate ticket and we can look into possible solutions!

**Comment by vlerenc** [comment]
OK, thank you. Given that we make heavy use of security groups for service instance isolation it is kind of an issue for the ops colleagues managing the larger landscapes. Cheers, Vedran

**Comment by vlerenc** [comment]
Created #892.
